### PR TITLE
Add RFC for start_trajectory service simplification.

### PR DESCRIPTION
**Pull request by MichaelGrupp**
[rendered draft](https://github.com/MichaelGrupp/rfcs/blob/refactor-start-trajectory/text/0028-simplify-ros-start-trajectory.md)

### CODE REVIEW

File: text/0000-simplify-ros-start-trajectory.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+It should be possible to start a trajectory without initial pose, just as now when the initial pose argument isn't given to the start_trajectory executable.
+
+Unlike e.g. protobuf, there's no way to define a part of a message definition as optional in ROS.
+`to_trajectory_id` would be zero initialized if unspecified, which is a valid trajectory ID.
+Maybe it's possible to find a workaround:
+
+* hide the pose in an array that can have zero elements (ugly)
+* two services: `/start_trajectory_with_initial_pose`, `start_trajectory`
+* ...?
```
Comment (MichaelGrupp) [inline text/0000-simplify-ros-start-trajectory.md]: To do: starting without initial pose == using an identity pose?

File: text/0000-simplify-ros-start-trajectory.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+## Discussion Points
+[discussion]: #discussion
+
+### Optional initial pose (TODO)
+
+It should be possible to start a trajectory without initial pose, just as now when the initial pose argument isn't given to the start_trajectory executable.
+
+Unlike e.g. protobuf, there's no way to define a part of a message definition as optional in ROS.
+`to_trajectory_id` would be zero initialized if unspecified, which is a valid trajectory ID.
+Maybe it's possible to find a workaround:
```
Comment (MichaelGrupp) [inline text/0000-simplify-ros-start-trajectory.md]: Consensus from open house meeting: add boolean `use_initial_pose` - easy to understand and easy to use

File: text/0000-simplify-ros-start-trajectory.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+* hide the pose in an array that can have zero elements (ugly)
+* two services: `/start_trajectory_with_initial_pose`, `start_trajectory`
+* ...?
+
+### Keep time parameter?
+
+The time parameter of the initial pose is a bit of an obscure feature that might confuse people - should we keep it in the ROS service? See [cartographer/pose_graph.h](https://github.com/googlecartographer/cartographer/blob/bdb6f2db4a1f98484b222d61abceab8adb74dfd1/cartographer/mapping/pose_graph.h#L138)
```
Comment (MichaelGrupp) [inline text/0000-simplify-ros-start-trajectory.md]: Consensus from open house meeting: don't expose in ROS service

File: text/0000-simplify-ros-start-trajectory.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+It hides the implementation details away from the user/client with a proper interface that has only 3 simple options.
+
+## Approach
+[approach]: #approach
+
+Change the service definition so that clients only need to depend on message packages (`cartographer_ros_msgs`, `geometry_msgs`, `std_msgs`), e.g.:
+
+```
+string configuration_directory
+string configuration_basename
```
Comment (MichaelGrupp) [inline text/0000-simplify-ros-start-trajectory.md]: As mentioned by @ojura in the open house meeting, this requires that the configuration lies on the host filesystem of the Cartographer node. Binary proto allowed to transfer options from a client in the network.

We agreed that we will drop this possibility for the sake of simplicity, which will benefit more users. Also, client/server scenarios are better covered by Cartographer gRPC anyway.

**Review by MichaelGrupp — COMMENTED** [review]
@ojura @gaschler Added quick notes from the meeting. Will update the doc soon.

**Comment by MichaelGrupp** [comment]
@gaschler Ready to merge from my side, PTAL.

**Comment by gaschler** [comment]
@ojura I remember you had quite

File: text/0000-simplify-ros-start-trajectory.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+cartographer_ros_msgs/StatusResponse status
+int32 trajectory_id
+```
+
+Then, change the service handler in `node.cc` to do what `start_trajectory_main.cc` has been doing so far (essentially: loading the configuration).
+
+As a nice side effect we can get rid of the following:
+
+* `start_trajectory_main.cc`
+* `TrajectoryOptions.msg`
```
Comment (gaschler) [inline text/0000-simplify-ros-start-trajectory.md]: Yeah!

File: text/0000-simplify-ros-start-trajectory.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+### Advantages of passing options in the message?
+
+As mentioned by @ojura in the open house meeting, the proposed approach requires that the configuration lies on the host filesystem of the Cartographer node.
+Binary proto allowed to transfer options from a client in the network.
+
+**Decision**:
+We agreed that we will drop this possibility for the sake of simplicity, which will benefit more users.
+Also, client/server scenarios are better covered by Cartographer gRPC anyway.
+
```
Comment (gaschler) [inline text/0000-simplify-ros-start-trajectory.md]: Are we losing an important feature about topic remapping?
In my eyes, the node (server) should know about the sensor topics available and needs to be started with topic remappings.
Who makes the "start_trajectory" service call has less insight what topics are available.
Removing the SensorTopics.msg, are we losing an important feature here that cannot be achieved by the new design?

**Review by gaschler — COMMENTED** [review]
Generally agree.
I'm happy how the interface becomes simpler.

(Sorry for closing this PR accidentally.)

File: text/0000-simplify-ros-start-trajectory.md (hunk 6 above)
Comment (MichaelGrupp) [reply text/0000-simplify-ros-start-trajectory.md]: I'm not sure I completely understand what you mean.

The remapping code in `start_trajectory_main` was added by @ojura in https://github.com/googlecartographer/cartographer_ros/pull/584. I think if we remove this executable and put all the logic in the node then we can just use sth similar to [StartTrajectoryWithDefaultTopics](https://github.com/googlecartographer/cartographer_ros/blob/41c74de649fca92b55c66ab0f2ee26be1c84ce69/cartographer_ros/cartographer_ros/node.h#L79) and remappings can be specified in the launch file of the node.

**Review by gaschler — APPROVED** [review]

**Review by cschuet — APPROVED** [review]

**Comment by MichaelGrupp** [comment]
Can we merge this? Then we can start working on this.

**Comment by gaschler** [comment]
Thank you, the RFC is approved, there is no open discussion.
Please rename the md file, then I can merge.

**Comment by MichaelGrupp** [comment]
Done, it's 0028 now.

**Comment by gaschler** [comment]
@MichaelGrupp Thank you, looking forward to the simplification

**Comment by MichaelGrupp** [comment]
:tada:

**Pull request merged**